File: yukicoder/tyama_yukicoder109.py
#!/usr/bin/python
for _ in range(input()):
	n,m=map(int,raw_input().split())
	a=0
	b=1
	while b*-~b<=m:
		b+=1
		if m%b<1:
			if n<b+m/b:
				a=1
				while n:
					a=a*n%m
					n-=1
			break
	else:
		if n<m:
			b=1
			n+=1
			while n<m:
				b=b*n%m
				n+=1
			a=(m-1)*pow(b,max(n-2,0),m)%m
	print(a)

File: yukicoder/tyama_yukicoder109.rb
#!/usr/bin/ruby
gets;$<.map{|e|	
	n,m=e.split.map &:to_i
	a=0
	if (2..Math.sqrt(m)).each{|b|
		(n.downto(a=1){|n|a=a*n%m}if n<b+m/b;break) if m%b<1
	}&&n<m
		(n+1..m-b=1).each{|n|b=b*n%m};a=1+n=m-2;(a=a*b%m if n%2>0;b=b*b%m;n/=2)while n>0
	end
	p a
}